File: implementations/go/oshash.go
// OpenSubtitles Hash (OSHash) - Go implementation.
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"os"
)

const ChunkSize = 65536

func computeHash(file *os.File) (uint64, error) {
	fi, err := file.Stat()
	if err != nil {
		return 0, err
	}
	fsize := fi.Size()
	if fsize < ChunkSize*2 {
		return 0, fmt.Errorf("file too small: %d bytes", fsize)
	}

	buf := make([]byte, ChunkSize*2)

	// Read first 64KB
	if _, err := file.ReadAt(buf[:ChunkSize], 0); err != nil {
		return 0, err
	}
	// Read last 64KB
	if _, err := file.ReadAt(buf[ChunkSize:], fsize-ChunkSize); err != nil {
		return 0, err
	}

	var nums [ChunkSize * 2 / 8]uint64
	reader := bytes.NewReader(buf)
	if err := binary.Read(reader, binary.LittleEndian, &nums); err != nil {
		return 0, err
	}

	var hash uint64
	for _, num := range nums {
		hash += num
	}

	return hash + uint64(fsize), nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <file>\n", os.Args[0])
		os.Exit(1)
	}

	file, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	hash, err := computeHash(file)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("%016x\n", hash)
}

File: implementations/nim/oshash.nim
# OpenSubtitles Hash (OSHash) - Nim implementation.
import os, strformat, strutils

const ChunkSize = 65536

proc computeHash(filepath: string): string =
  let fsize = getFileSize(filepath)
  if fsize < ChunkSize * 2:
    raise newException(ValueError, "File too small: " & $fsize & " bytes")

  var hash: uint64 = uint64(fsize)
  var f = open(filepath, fmRead)
  defer: f.close()

  var buf: array[8, uint8]

  # Hash first 64KB
  for i in 0 ..< (ChunkSize div 8):
    discard f.readBytes(buf, 0, 8)
    var val: uint64 = 0
    for j in 0..7:
      val = val or (uint64(buf[j]) shl (j * 8))
    hash = hash + val

  # Hash last 64KB
  f.setFilePos(max(0, fsize - ChunkSize))
  for i in 0 ..< (ChunkSize div 8):
    discard f.readBytes(buf, 0, 8)
    var val: uint64 = 0
    for j in 0..7:
      val = val or (uint64(buf[j]) shl (j * 8))
    hash = hash + val

  return toLowerAscii(fmt"{hash:016x}")

when isMainModule:
  let params = commandLineParams()
  if params.len < 1:
    stderr.writeLine("Usage: oshash <file>")
    quit(1)
  echo computeHash(params[0])
